Splitter refactoring

arvoelke opened the pull request:
Closes #80. Refactors the splitter logic for adding / removing connections to occur within the builder. This consolidates the connection logic into one place (`builder/connection.py`) so that backend-relevant logic and objects are manipulated at the same level of abstraction.

Overarching ideas:
 - The splitter returns "directives" (i.e., sets of instructions) to guide the builder.
 - The builder fills in all three "sub-models" (`self`, `host`, and `host_pre`) in parallel, by delegating the objects to the appropriate builder methods.

Solves the following bugs:
 - https://github.com/nengo/nengo-loihi/issues/205 - Splitter handles sliced probes.
 - https://github.com/nengo/nengo-loihi/issues/206 - Passthrough removal handles sliced probes.
 - https://github.com/nengo/nengo-loihi/issues/207 - Decoder cache is not globally disabled.
 - https://github.com/nengo/nengo-loihi/issues/208 - Improved chip -> chip learning signal message.
 - https://github.com/nengo/nengo-loihi/issues/209 - Improved chip -> chip error signal message.
 - https://github.com/nengo/nengo-loihi/issues/211 - Splitter won't mutate original network.

Random improvements:
 - ~~`split_pre_from_host` is no longer `O(n^2)`~~ (This was a misunderstanding on my part, fueled by the fact the variable named "queue" was actually a stack -- it is now actually a queue)
 - `__contains__` is no longer `O(n)`

Bugs that have been identified that are outside the scope of this PR:
 - https://github.com/nengo/nengo-loihi/issues/210 - Passthrough connections aren’t seeded, and the order is non-deterministic.
 - https://github.com/nengo/nengo-loihi/issues/212 - Passthrough removal is too aggressive.
 - https://github.com/nengo/nengo-loihi/issues/213 - Passthrough removal is broken.
 - https://github.com/nengo/nengo-loihi/issues/214 - `precompute=True` could be smarter.

Existing issues that should now be more readily achievable:
 - #163 - Passing instance config parameters to builder
 - #154 - `sim.data` does not contain chip <-> host connections
 - #145 - Efficient passthrough nodes
 - #118 - Indicate which parts of a model are running on/off chip
 - #4 - Expose `params` and handle splitter-created objects

Other future work:
 - Ability to pass in `network=None`.
 - Change `assert self.model.dt == dt` to an error and then unit test (https://github.com/nengo/nengo/issues/1517).

arvoelke commented:
FYI, the doc build gave me:
```
/home/travis/build/nengo/nengo-loihi/docs/examples.rst.rst:33:broken link: https://github.com/nengo/nengo-examples/blob/master/loihi-dl/cifar10_convnet.ipynb (403 Client Error: Forbidden for url: https://github.com/nengo/nengo-examples/blob/master/loihi-dl/cifar10_convnet.ipynb)
```

hunse commented on nengo_loihi/builder/tests/test_connection.py:
```diff
@@ -0,0 +1,67 @@
+@pytest.mark.parametrize("pre_dims", [1, 3])
+@pytest.mark.parametrize("post_dims", [1, 3])
+@pytest.mark.parametrize("learn", [True, False])
+@pytest.mark.parametrize("use_solver", [True, False])
```
This is 16 tests. Do we need every possible combination here? They don't take long on the emulator, but I'm worried they could add more time than necessary on the chip.

arvoelke replied:
This unit test was copied from the old splitter tests. Perhaps this should be added to a card to investigate/prune the unit tests for CI?

hunse replied:
Dan already did some of this not too long ago. If this is from before, it's probably already fairly efficient. We can leave it for now, and perhaps in a few months we'll be revisiting tests again and can do some pruning.

hunse commented on nengo_loihi/simulator.py:
```diff
@@ -306,64 +305,65 @@ def __init__(  # noqa: C901
-
-            # Build the network into the model
-            self.model.build(network)
+        if network is None:
```
Why stop allowing `network == None`?

arvoelke replied:
We had some discussion about this in nengo slack. This wasn't being used (or tested) anywhere, and---prior to the refactoring---didn't make a whole lot of sense, because, if we look at these lines in the original version:

https://github.com/nengo/nengo-loihi/blob/da2053ff5847fa8c68b4a4d30378c2cead1edb58/nengo_loihi/simulator.py#L313-L366

there are a large number of attributes added to `self` within this block that are required later. So a user would need to do a few very special things in order to get this to work, which might easily change. Exposing this kind of untested surface area for no good reason seemed like a bad idea to me. @tbekolay agreed that it made sense to disallow for now.

After this refactoring is complete, it could make sense to add back in. Currently I think the need for the attribute `self.model.splitter_directive` being filled in is really the only thing holding this back for now. I'd love to have a better way of organizing the way that model parameters are filled in within the simulator / builder.

hunse replied:
I agree, there's not a whole lot of demand for it and it's not well-tested. It is something that we allow in the reference backend. I think going forward, we need to discuss whether it's something we want to support, and if so, add a `nengo` test for it to make sure backends support it. I've made an issue nengo/nengo#1519.

hunse commented on nengo_loihi/splitter.py:
```diff
@@ -1,543 +1,141 @@
-    # --- Step 5: place probes
-    place_probes(networks)
+            self.passthrough_directive = PassthroughDirective(
+                set(), set(), set())
```
Seems a bit strange to import both `convert_passthroughs` and `PassthroughDirective`. I would pick one as the interface, and have it somehow able to return and empty directive in the case that `remove_passthrough == False`.

arvoelke replied:
Yeah it's also strange that I import `is_passthrough` and have to manually filter things before calling the function. But right now all this weirdness/friction is limited to 6 lines of code, and I have a feeling all of it will change when we redo the passthrough removal. Would it be better to define an `EmptyPassthroughDirective` within passthrough.py and then import+use that?

hunse replied:
I don't think that would make much of a difference. That's a fair point that it's only 6 lines and might change, and there's not an obvious way to do it better.

hunse commented on nengo_loihi/splitter.py:
```diff
@@ -1,543 +1,141 @@
-    conns.extend(add_conns)
+    def _preclosure(self):  # noqa: C901
+        # performs a transitive closure on all host objects that
+        # send data to the chip
```
What does this mean? Also, I'd make this a brief docstring.

arvoelke replied:
https://en.wikipedia.org/wiki/Transitive_closure#In_graph_theory

Will add docstring.

hunse commented on nengo_loihi/splitter.py:
```diff
@@ -1,543 +1,141 @@
-                    and networks.location(obj) == "host"):
-                if isinstance(obj, HostReceiveNode):
+        # initialize queue with the pre objects on
+        # host -> chip connections
```
Why are all `pre` objects on host->chip connections precomputable? Couldn't you have something like node->ensemble->node->ensemble->node (where nodes are on host and ensembles are on chip) where the middle node would be the `pre` of a host->chip connection but not be precomputable?

arvoelke replied:
Great catch... I'm amazed that none of the tests caught this. I pushed a commit 0cbe29b which tests this and makes the fix (namely, using the right set of connections to form the adjacency lists).

hunse replied:
This still appears to assume that all `pre` objects on host->chip connections are precomputable. This might be true in some cases, like in the test you added where the passthrough nodes can be removed, but what if those weren't passthroughs (i.e. if all nodes in that test computed functions)? I think in general, you can have networks where some `pre` objects of connections are not precomputable.

arvoelke replied:
Yep.. this is the same assumption that passthrough removal has always been making. If anything in the transitive closure of objects that send to host (after passthrough removal) also receives data from the host then a `BuildError` is raised. That is, the example that you suggested currently (and previously) raises an error. There is a unit test for this (`test_split_precompute_loop_error`). Note I didn't change the algorithm here. I just updated it to be a bit clearer IMO and more efficient (it used to be `O(n^2)`). I talked about this in slack but there was no interest in making this smarter at the moment. I've summarized this alongside a similar issue now in #214.

The difference that you initially pointed out stemmed from the fact that previously the passthrough removal would be physically applied to the `SplitNetworks` object before computing the closure. In this PR the list of connections weren't changed, so I added the above commit to make sure it's using the same list as it had been in the original splitter.

By the way we should remove "*Various improvements to passthrough removal, including not removing useful passthrough nodes.*" from the commit message. This was squashed from one of the other commits that was fixing a mistake I had previously made. There aren't any improvements to passthrough removal in this PR apart from making it `O(n)` instead of `O(n^2)`.

hunse replied:
Okay, maybe some comments in the code to that effect would be helpful. I can draft them.

That's good to know about that part of the commit message. I had wondered about that when I put it in, because I couldn't see where the stuff about "not removing useful passthrough nodes" happened, but I didn't want to get rid of it until I checked with you. As for "various improvements", I was referring to renaming "offchip" to "ignore", supporting `ObjView`, and returning a named tuple instead of a tuple. I don't actually see that the passthrough removal algorithm changes at all.

arvoelke replied:
"I don't actually see that the passthrough removal algorithm changes at all." Which part are you referring to here? The `O(n^2) -> O(n)` part? This is just from avoiding the repeated list pops, which each take `O(n)` time, by instead advancing a `head` index.

tbekolay replied:
List pops in Python are O(1) (see [time complexity](https://wiki.python.org/moin/TimeComplexity))

arvoelke replied:
That's from an amortized analysis or average case analysis. Here we were always popping from the beginning of a list of size `O(n)` which is `O(n)` each pop. Quote from the same page:

> Internally, a list is represented as an array; the largest costs come from growing beyond the current allocation size (because everything must move), or from inserting or deleting somewhere near the beginning (because everything after that must move). If you need to add/remove at both ends, consider using a collections.deque instead.

Since lists are just arrays, `list.pop(0)` needs to shift the entire array each time.

tbekolay replied:
I thought we were doing `queue.pop()` ? Which pops from the end of the list

arvoelke replied:
Ah, you are right. The naming of the variable was a misnomer. Since we append to the last and pop from the last, it is not a queue -- it is a _stack_. It doesn't actually matter which one it is, but I was trying to keep things consistent with a queue. So either we can change it back to `pop` and rename it from `queue -> stack` or leave it as is. The ordering shouldn't matter, but should make it clear.

hunse replied:
This isn't the passthrough removal algorithm, though, this is the splitter. That's all I was trying to say, is that I didn't see any changes to passthrough removal.

arvoelke replied:
Apparently my brain has started interchanging the words "passthrough removal" and "splitter" as well as "chip" and "host". Ha ha .. :x You're right that I didn't change anything about passthrough removal other than fixing a bug that affected probed `ObjView`s.

arvoelke replied:
Pushed commit 433b489 and updated issue #214 according to the above conversation about precomputing.

arvoelke commented on nengo_loihi/builder/node.py:
```diff
@@ -1,7 +1,7 @@
 
 from nengo_loihi.builder.builder import Builder
-from nengo_loihi.inputs import ChipReceiveNode, SpikeInput
+from nengo_loihi.builder.inputs import ChipReceiveNode, SpikeInput
```
Yeah I was thinking about separating out that file as well. Currently `SpikeInput` is the only one that is used from outside the `/builder/` directory. In particular, `hardware/builder.py` and `emulator/tests/test_interface.py` import `SpikeInput`, and the rest are contained within the `/builder/`.

```$ grep "from nengo_loihi.builder.inputs" . -r
./nengo_loihi/builder/node.py:from nengo_loihi.builder.inputs import ChipReceiveNode, SpikeInput
./nengo_loihi/builder/connection.py:from nengo_loihi.builder.inputs import (
./nengo_loihi/builder/builder.py:from nengo_loihi.builder.inputs import LoihiInput
./nengo_loihi/builder/tests/test_inputs.py:from nengo_loihi.builder.inputs import ChipReceiveNode, PESModulatoryTarget
./nengo_loihi/conv.py:from nengo_loihi.builder.inputs import ChipReceiveNeurons, LoihiInput
./nengo_loihi/tests/test_simulator.py:from nengo_loihi.builder.inputs import SpikeInput
./nengo_loihi/tests/test_block.py:from nengo_loihi.builder.inputs import SpikeInput
./nengo_loihi/hardware/builder.py:from nengo_loihi.builder.inputs import SpikeInput
./nengo_loihi/emulator/tests/test_interface.py:from nengo_loihi.builder.inputs import SpikeInput
```

One exception is `conv.py`, which imports `ChipReceiveNeurons` and `LoihiInput` to use within `build_conv2d_connection`. However, there is a TODO in `builder/connection.py::build_chip_connection` next to the call to `build_conv2d_connection` that says "TODO: integrate these into the same function" and so this is supposed to be in `/builder/` anyway.

What if I were to move `LoihiInput` and `SpikeInput` to `block.py`?

hunse replied:
If we're leaving the builder-specific ones in `builder/`, then, `LoihiInput` and `SpikeInput` *could* go in `block.py`. I'm trying to predict whether we'll have more in the near future, or not, because that would be the main thing that would motivate having `inputs.py`. The main potential one I can see on the horizon is `DVSInput`.

I think I'd vote to keep `LoihiInput` and `SpikeInput` in `inputs.py`, and move the others to `builder/inputs.py`.

Also, we should move the conv connection builder to `builder/connection.py` I think.

arvoelke commented:
I ran @tcstewar's benchmark notebook with:
```python
conditions = [
    'nengo',                                         # the normal nengo backend
    'nengo_loihi(precompute=False, target="loihi")', # using Loihi itself
    'nengo_loihi(precompute=True, target="loihi")',  # using Loihi itself with precompute
]
```
on both `master` and this branch.

On this branch, `precompute=False` and `precompute=True` gave identical accuracy (but with different speeds) across all benchmarks and so I added a simple unit test (45eced7) to verify this. The learning benchmarks aren't supposed to work with `precompute=True` (#214) and so I added a commit (433b489) to improve the error message.

On master, `precompute=False` and `precompute=True` gave different results for `['CircularConvolution', 'MatrixMultiply', 'SPASequence']`, which I presume has something to do with some interaction between `remove_passthroughs=True` and `precompute` with the seeding of the network in the old splitter code (also see issue #210).

arvoelke commented:
Pushed two more commits (1711044, 0ca11d8) to move `LoihiInput` and `SpikeInput` from `builder/inputs.py` back into `inputs.py` and move `build_conv2d_connection` from `conv.py` to `builder/connections.py`. This should now address all your comments @hunse.

arvoelke commented:
Added a commit (6c201e0) that increases test coverage since I touched the lines of some un-tested convolution code.

hunse commented on nengo_loihi/builder/connection.py:
```diff
@@ -282,6 +282,93 @@ def build_host_to_learning_rule(model, conn):
     nengo_model.build(pre2send)
 
 
+def build_conv2d_connection(model, conn):
```
I would probably put it after `build_chip_connection`, since it's a subset of that, but not a big deal either way.

hunse commented on nengo_loihi/builder/builder.py:
```diff
@@ -167,8 +189,38 @@ def add_block(self, block):
         assert block not in self.blocks
         self.blocks[block] = len(self.blocks)
 
+    def delegate(self, obj):
```
It would be helpful to have a docstring here.

hunse commented on nengo_loihi/builder/connection.py:
```diff
@@ -1,24 +1,372 @@
+
 import nengo
-from nengo import Ensemble, Connection, Node
+from nengo import Ensemble, Connection, Node, Probe as NengoProbe
```
I would maybe just `import nengo` and use `nengo.` for all of these.

hunse commented:
Just a few minor comments. Otherwise everything LGTM. Thanks for doing all those tests against the benchmarks.

arvoelke commented:
I added a couple very minor commits for coverage / sanity. I also rebased onto master again to pick up the testing fixes from the current/rates sync-up.

> 5694c27 Is Split a good name? Or should we stick with Splitter, SplitterDirective, or some other option?
15890be Similarly, is PassthroughSplit a good name?
ea6d52e Should this be a separate PR or is it OK to include here?

LGTM!

> 16e322a Is get_seed something we want built into Model? And if so, do we do it now or make a separate issue for it?

I think it's fine as is. A bunch of the weirdness with seeding can be resolved when we eventually drop support for nengo<3. This could end up simplifying a bunch of this code.

> f64b102 Are BuildErrors sufficient here or do we need to make a new NengoException subclass?

This LGTM, but we should make a card to go through the rest of the code base? e..g, There are a bunch of `ValueError` exceptions in `hardware/builder.py`.

arvoelke commented:
I just noticed that the docstring simulator says the default for `precompute=True`. We should fix that here IMO.

tbekolay commented:
> We should fix that here IMO.

OK, will do. Running into some issues getting the hardware build to pass so I'll be doing at least one more push anyhow.
